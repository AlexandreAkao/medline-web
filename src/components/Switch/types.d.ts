interface ISwitchContainer {
  isLeft: boolean;
}

interface ISwitchOption {
  label: string;
  value: string | number;
}

interface ISwitchProps {
  optionLeft: ISwitchOption;
  optionRight: ISwitchOption;
  onChange?: (option: ISwitchOption) => void;
}
