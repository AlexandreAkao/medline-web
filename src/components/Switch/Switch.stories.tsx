import { ComponentStory, ComponentMeta } from '@storybook/react';

import Switch from 'components/Switch';

export default {
  title: 'Switch',
  args: {},
  component: Switch,
} as ComponentMeta<typeof Switch>;

const Template: ComponentStory<typeof Switch> = args => <Switch {...args} />;

export const MaxWidth = Template.bind({});
MaxWidth.args = {
  optionLeft: { label: 'Atestado', value: 'left' },
  optionRight: { label: 'Receita', value: 'right' },
  onChange: option => console.log(option),
};

export const MediumWidth = Template.bind({});
MediumWidth.args = {
  optionLeft: { label: 'Atestado', value: 'left' },
  optionRight: { label: 'Receita', value: 'right' },
};
MediumWidth.decorators = [
  Story => (
    <main style={{ width: '50vw' }}>
      <Story />
    </main>
  ),
];

export const SmallWidth = Template.bind({});
SmallWidth.args = {
  optionLeft: { label: 'Atestado', value: 'left' },
  optionRight: { label: 'Receita', value: 'right' },
};
SmallWidth.decorators = [
  Story => (
    <main style={{ width: '20vw' }}>
      <Story />
    </main>
  ),
];
