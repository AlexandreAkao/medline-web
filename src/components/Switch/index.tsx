import { useState } from 'react';

import {
  SwitchButton,
  SwitchContainer,
  SwitchInput,
  SwitchLabel,
  SwitchOption,
} from 'components/Switch/styles';

function Switch({ optionLeft, optionRight, onChange }: ISwitchProps) {
  const [isRight, setIsRight] = useState(false);

  const handleOnChange = () => {
    onChange && onChange(!isRight ? optionRight : optionLeft);
    setIsRight(prev => !prev);
  };

  return (
    <SwitchContainer>
      <SwitchInput
        id="react-switch-new"
        name="react-switch-new"
        type="checkbox"
        data-testid="test-switch"
        checked={isRight}
        onChange={handleOnChange}
      />
      <SwitchLabel htmlFor="react-switch-new">
        <SwitchOption>
          <span>{optionLeft.label}</span>
          <span>{optionRight.label}</span>
        </SwitchOption>
        <SwitchButton />
      </SwitchLabel>
    </SwitchContainer>
  );
}

export default Switch;
