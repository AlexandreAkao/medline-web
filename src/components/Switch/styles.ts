import styled from 'styled-components';

import colors from 'styles/colors';

export const SwitchContainer = styled.div`
  display: flex;

  border: 1.5px solid ${colors.black.soft};
  border-radius: 40px;
  background-color: ${colors.white.normal};
  user-select: none;

  cursor: pointer;
`;

export const SwitchOption = styled.div`
  display: flex;
  width: 100%;
  height: 100%;
  align-items: center;
  justify-content: center;

  span {
    padding: 0 20px;
    font-size: 1.125rem;
    font-weight: 700;
    flex: 1 1 0;
    z-index: 1;
    text-align: center;
    color: ${colors.black.soft};
  }

  span:first-child {
    color: ${colors.white.normal};
  }
`;

export const SwitchButton = styled.span`
  content: '';
  position: absolute;
  top: 2px;
  left: 2px;
  width: 50%;
  height: 36px;
  border-radius: 45px;
  transition: 0.2s;
  background-color: ${colors.primary};
  box-shadow: 0 0 2px 0 rgba(10, 10, 10, 0.3);
`;

export const SwitchLabel = styled.label`
  display: flex;
  align-items: center;
  justify-content: space-between;
  height: 40px;
  width: 100%;
  background-color: ${colors.white.normal};
  border-radius: 100px;
  position: relative;
  transition: background-color 0.2s;
  cursor: pointer;

  :active ${SwitchButton} {
    width: 55%;
  }
`;

export const SwitchInput = styled.input`
  height: 0;
  width: 0;
  visibility: hidden;

  :checked + ${SwitchLabel} {
    ${SwitchButton} {
      left: calc(100% - 2px);
      transform: translateX(-100%);
    }

    ${SwitchOption} {
      span:last-child {
        color: ${colors.white.normal};
      }
      span:first-child {
        color: ${colors.black.soft};
      }
    }
  }
`;
