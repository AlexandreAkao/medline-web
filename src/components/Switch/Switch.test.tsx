import { render, screen } from '@testing-library/react';
import { composeStories } from '@storybook/testing-react';
import userEvent from '@testing-library/user-event';
import { vi } from 'vitest';

import * as stories from 'components/Switch/Switch.stories';
import Switch from 'components/Switch';

const { MaxWidth, MediumWidth, SmallWidth } = composeStories(stories);

describe('Switch', () => {
  const renderComponent = (customProps: Partial<ISwitchProps> = {}) => {
    const props: ISwitchProps = {
      optionLeft: { label: 'left', value: 'left' },
      optionRight: { label: 'right', value: 'right' },
      ...customProps,
    };
    return render(<Switch {...props} />);
  };

  it('should match with snapshot', () => {
    const { container } = renderComponent();
    expect(container).toMatchSnapshot();
  });

  it('should render a switch with max width', () => {
    render(<MaxWidth />);
    const leftElement = screen.getByText(/Atestado/i);
    const rightElement = screen.getByText(/Receita/i);

    expect(leftElement).toBeInTheDocument();
    expect(rightElement).toBeInTheDocument();
  });

  it('should render a switch with medium width', () => {
    render(<MediumWidth />);

    const mainElement = screen.getByRole('main');
    const leftElement = screen.getByText(/Atestado/i);
    const rightElement = screen.getByText(/Receita/i);

    expect(leftElement).toBeInTheDocument();
    expect(rightElement).toBeInTheDocument();
    expect(mainElement).toHaveStyle({
      width: '50vw',
    });
  });

  it('should render a switch with small width', () => {
    render(<SmallWidth />);

    const mainElement = screen.getByRole('main');
    const leftElement = screen.getByText(/Atestado/i);
    const rightElement = screen.getByText(/Receita/i);

    expect(leftElement).toBeInTheDocument();
    expect(rightElement).toBeInTheDocument();
    expect(mainElement).toHaveStyle({
      width: '20vw',
    });
  });

  it('should render a switch with small width', () => {
    render(<SmallWidth />);

    const mainElement = screen.getByRole('main');
    const leftElement = screen.getByText(/Atestado/i);
    const rightElement = screen.getByText(/Receita/i);

    expect(leftElement).toBeInTheDocument();
    expect(rightElement).toBeInTheDocument();
    expect(mainElement).toHaveStyle({
      width: '20vw',
    });
  });

  it('should trigger on change when clicked', async () => {
    const onChange = vi.fn();
    renderComponent({ onChange });

    const switchElement = screen.getByTestId('test-switch');
    const leftElement = screen.getByText(/left/i);
    const rightElement = screen.getByText(/right/i);

    await userEvent.click(switchElement);

    expect(leftElement).toBeInTheDocument();
    expect(rightElement).toBeInTheDocument();
    expect(switchElement).toBeChecked();
    expect(onChange).toHaveBeenCalledWith({ label: 'right', value: 'right' });

    await userEvent.click(switchElement);

    expect(onChange).toHaveBeenCalledWith({ label: 'left', value: 'left' });
  });
});
