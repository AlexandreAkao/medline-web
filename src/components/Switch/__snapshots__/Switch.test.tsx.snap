// Vitest Snapshot v1

exports[`Switch > should match with snapshot 1`] = `
<div>
  <div
    class="sc-bczRLJ kCIQoJ"
  >
    <input
      class="sc-eCYdqJ iXaHfl"
      data-testid="test-switch"
      id="react-switch-new"
      name="react-switch-new"
      type="checkbox"
    />
    <label
      class="sc-hKMtZM bBCKhs"
      for="react-switch-new"
    >
      <div
        class="sc-gsnTZi gkDFwC"
      >
        <span>
          left
        </span>
        <span>
          right
        </span>
      </div>
      <span
        class="sc-dkzDqf jHkIlX"
      />
    </label>
  </div>
</div>
`;
